### PR TITLE
[SPARK-46403][SQL] Decode parquet binary with getBytesUnsafe method

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

Now spark will get a parquet binary object with getBytes() method.

The **Binary.getBytes()** method will always make a new copy of the internal bytes.

We can use **Binary.getBytesUnsafe()** method  in  https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/io/api/Binary.java#L55-L62 to get the cached bytes if it has already been called getBytes() and has the cached bytes.

Local benchmark, before this PR:
```
OpenJDK 64-Bit Server VM 17.0.6+10-LTS on Mac OS X 13.6
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Parquet dictionary:                       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Read binary dictionary                            18919          19449         393          5.5         180.4       1.0X
```
after this PR:
```
OpenJDK 64-Bit Server VM 17.0.6+10-LTS on Mac OS X 13.6
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Parquet dictionary:                       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Read binary dictionary                            10135          10602         337         10.3          96.7       1.0X

```

### Why are the changes needed?

Optimize parquet reader.
Before this PR:
image

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Local test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by wankunde** [comment]
Local benchmark code

```scala
object ParquetDictionaryDecoderBenchmark extends SqlBasedBenchmark {

  import spark.implicits._

  private val N = 100 * 1024 * 1024
  private val NUMBER_OF_ITER = 10


  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    withTempPath { tempDir =>
      val outputPath = tempDir.getCanonicalPath
      val df: DataFrame = spark
        .range(1, N, 1, 1)
        .map(id => ("ABCDEFG", "HIJKLMN", "OPQRST", "UVWXYZ"))
        .toDF("a", "b", "c", "d")
      df.write.mode(SaveMode.Overwrite).parquet(outputPath)
      val benchmark = new Benchmark("Parquet dictionary", N, NUMBER_OF_ITER, output = output)
        benchmark.addCase("Read binary dictionary") { _ =>
          spark.read.parquet(outputPath).noop()
        }
      benchmark.run()
    }
  }
}
```

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetDictionary.java [hunk 1]
```diff
@@ -70,7 +70,7 @@ public byte[] decodeToBinary(int id) {
       long signed = dictionary.decodeToLong(id);
       return new BigInteger(Long.toUnsignedString(signed)).toByteArray();
     } else {
-      return dictionary.decodeToBinary(id).getBytes();
+      return dictionary.decodeToBinary(id).getBytesUnsafe();
```
[the remark this answers is not in the document]
Comment (wankunde) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetDictionary.java:73]: The comment of `getBytesUnsafe()` method:
```
  /**
   * Variant of getBytes() that avoids copying backing data structure by returning
   * backing byte[] of the Binary. Do not modify backing byte[] unless you know what
   * you are doing.
   *
   * @return backing byte[] of correct size, with an offset of 0, if possible, else returns result of getBytes()
   */
  public abstract byte[] getBytesUnsafe();
```
Spark will copy those cached bytes before using them, so I think it's safe for spark.

**Review by yaooqinn — APPROVED** [review]
LGTM. I have checked that Hive also uses `getBytesUnsafe`.

**Comment by yaooqinn** [comment]
also cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by beliefer** [comment]
Merged to master
@wankunde @yaooqinn @cloud-fan Thank you.

**Pull request closed without merging**